watchers/statefulset_handler.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 Datadog, Inc.
+package watchers
+
+import (
+	context "context"
+	"time"
+
+	chaosv1beta1 "github.com/DataDog/chaos-controller/api/v1beta1"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type StatefulSetHandler struct {
+	Client client.Client
+	log    *zap.SugaredLogger
+}
+
+func NewStatefulSetHandler(client client.Client, logger *zap.SugaredLogger) StatefulSetHandler {
+	return StatefulSetHandler{
+		Client: client,
+		log:    logger,
+	}
+}
+
+// OnAdd is a handler function for the add of a statefulset
+func (h StatefulSetHandler) OnAdd(obj interface{}) {
+	statefulset, ok := obj.(*appsv1.StatefulSet)
+
+	// If the object is not a statefulset, do nothing
+	if !ok {
+		return
+	}
+
+	// If statefulset doesn't have associated disruption rollout, do nothing
+	hasDisruptionRollout, err := h.HasAssociatedDisruptionRollout(statefulset)
+	if err != nil {
+		return
+	}
+
+	if !hasDisruptionRollout {
+		return
+	}
+
+	initContainersHash, containersHash, err := HashPodSpec(&statefulset.Spec.Template.Spec)
+	if err != nil {
+		return
+	}
+
+	err = h.UpdateDisruptionRolloutStatus(statefulset, initContainersHash, containersHash)
+	if err != nil {
+		return
+	}
+}
+
+// OnUpdate is a handler function for the update of a statefulset
+func (h StatefulSetHandler) OnUpdate(oldObj, newObj interface{}) {
+	// Convert oldObj and newObj to Deployment objects
+	oldStatefulSet, okOldStatefulSet := oldObj.(*appsv1.StatefulSet)
+	newStatefulSet, okNewStatefulSet := newObj.(*appsv1.StatefulSet)
+
+	// If both old and new are not statefulsets, do nothing
+	if !okOldStatefulSet || !okNewStatefulSet {
+		return
+	}
+
+	// If statefulset doesn't have associated disruption rollout, do nothing
+	hasDisruptionRollout, err := h.HasAssociatedDisruptionRollout(newStatefulSet)
+	if !hasDisruptionRollout || err != nil {
+		return
+	}
+
+	// If containers have't changed, do nothing
+	containersChanged, initContainersHash, containersHash, err := ContainersChanged(&oldStatefulSet.Spec.Template.Spec, &newStatefulSet.Spec.Template.Spec, h.log)
+	if !containersChanged || err != nil {
+		return
+	}
+
+	err = h.UpdateDisruptionRolloutStatus(newStatefulSet, initContainersHash, containersHash)
+	if err != nil {
+		return
+	}
+}
+
+// OnDelete is a handler function for the delete of a statefulset
+func (h StatefulSetHandler) OnDelete(_ interface{}) {
+	// Do nothing on delete event
+}
+
+func (h StatefulSetHandler) FetchAssociatedDisruptionRollouts(statefulset *appsv1.StatefulSet) (*chaosv1beta1.DisruptionRolloutList, error) {
+	indexedValue := "statefulset" + "-" + statefulset.Namespace + "-" + statefulset.Name
+
+	disruptionRollouts := &chaosv1beta1.DisruptionRolloutList{}
+	err := h.Client.List(context.Background(), disruptionRollouts, client.MatchingFields{"targetResource": indexedValue})
```

Comment:
This field selector is checking the spec? Let's leave a TODO here that it would be more efficient to use label selectors, though this would require a webhook to add those labels when new rollouts are created